### ISSUE TITLE
[Tracking Issue] Whole project review

### ISSUE THREAD

**Issue by flip1995** [Rust-GCC/cargo-gccrs#36]
This is a review of the binary library split branch: [`cleanup-error-handling`](https://github.com/Rust-GCC/cargo-gccrs/tree/cleanup-error-handling)

_First posted in https://github.com/Rust-GCC/cargo-gccrs/pull/21#pullrequestreview-700690940_, but I wanted to also track this outside of the PR.

I figured that this is now the right time to do a whole project review, and not only a review of #21. My suggestions don't have to be implemented all in one PR, but rather get implemented step by step. I tried to sort them from easy to hard to do.

---

- [x] `cfg!(release)`

What is this line supposed to do? https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/gccrs.rs#L23

---

- [x] `is_installed`

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/gccrs.rs#L43-L44

You can use the [`which`](https://crates.io/crates/which) crate for cross-platform support for this 

---

- [x] `serde` for config dump/parsing?

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L1-L4

What format is the config dump in and can you use serde for this?

---

- [x] Config dump/parsing impl

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L101-L102

A fallible display impl seems weird to me. You should rather implement a fallible constructor and then the `Display` trait. That way you can use the format machinery and get functions like `to_string` for free. 

I'm also not 100% sure what the option dumping is for, so please tell me if there is a reason why it has to be done that way.

---

- [x] multiple binaries?

I think you want to have 2 binaries here. One for the driver, that does all the argument conversion and other hard work, and one that just invokes the driver from the cargo command. But I may be misunderstanding something, see my comment https://github.com/Rust-GCC/cargo-gccrs/pull/21#discussion_r665144475

---

- [x] `GccrsArgs` not holding all arguments

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L171

I still don't quite get why it is necessary to return a `Vec<GccrsArgs>`. Just from the struct name and the documentation, I would assume that when I have a `GccrsArgs` instance, that I have all the arguments that I need. This is also the main thing that prevents the `GccrsArgs` from implementing things like the `TryFrom` trait and so on. 

---

- [x] `GccrsArgs` is doing too much

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L164-L168

I still think that with this, `GccrsArgs` is doing too much. One good principle of OOP is that one class should only do one thing. With that principle, `GccrsArgs` should only handle arguments, which includes parsing, converting, and then holding them. Now it handles arguments and does random other things like creating archives and in the future possibly even more.

This is the area of the current implementation that really needs a refactor IMHO. Apply the KISS principle as often as you can. I have trouble following what is done where, so the current impl isn't really simple to follow.

**Comment by flip1995** [comment]
I would address this in the following order:

1. Let's get the good parts of `cleanup-error-handling` merged first. Open a PR, that just adds `thiserror` to the library part (that is still kept as a module) and the binary part just calls `.expect` for now. Don't split lib/binary yet and don't use anyhow for now. 
2. Address the simple points 1-4
3. Address point 5
4. Address point 6-7

**Comment by CohenArthur** [comment]
>     * [ ]  `cfg!(release)`
> 
> 
> What is this line supposed to do?

Since `gccrs` cannot output some required information yet, we are faking it. Because I do not believe that this is valid solution, I've added this check to prevent the release build (and therefore what someone looking at crates.io might install) from working properly as it is not in a state to do so yet. For development builds, we get the fake output and can keep going. I am also okay with removing it, but thought it would be nice to mark that this is not the actual expected behavior.

>     * [ ]  `is_installed`
> 
> 
> https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/gccrs.rs#L43-L44
> 
> You can use the [`which`](https://crates.io/crates/which) crate for cross-platform support for this

Aaah, perfect! Thanks, I'll open an issue for this.


>     * [ ]  `serde` for config dump/parsing?
> 
> 
> https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L1-L4
> 
> What format is the config dump in and can you use serde for this?

I do not know if the config dump has a specific format. @philberty?
I've figured out the simple format used, but am not sure about how serializable it is. The current "parser" is very simple. I have never used a "new" format with serde so I wouldn't know how to proceed, but I am not against it.

>     * [ ]  Config dump/parsing impl
> 
> 
> https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L101-L102
> 
> A fallible display impl seems weird to me. You should rather implement a fallible constructor and then the `Display` trait. That way you can use the format machinery and get functions like `to_string` for free.

You're right. Thanks!

> I'm also not 100% sure what the option dumping is for, so please tell me if there is a reason why it has to be done that way.

When first launching `rustc`, `cargo` asks the compiler to print its configuration as well as a few other options such as the expected name of libraries, binaries etc. This was the first "milestone" to get through, you can find the expected output and associated issue [here](https://github.com/Rust-GCC/cargo-gccrs/issues/1)

>     * [ ]  multiple binaries?
> 
> 
> I think you want to have 2 binaries here. One for the driver, that does all the argument conversion and other hard work, and one that just invokes the driver from the cargo command. But I may be misunderstanding something, see my comment [#21 (comment)](https://github.com/Rust-GCC/cargo-gccrs/pull/21#discussion_r665144475)

I think it would be fine to have two binaries. Basically, what the current one is doing is the following:
- If launched as `cargo gccrs [build|run|test|...]`, spawn as a wrapper around `rustc` which will create the process `cargo-gccrs rustc ...`
- If lauched as `cargo-gccrs rustc`, act as a translator and produce `gccrs` options from `rustc` options.

Note the difference in the call with the extra dash. One invokes the cargo subcommand, while the other one invokes the actual binary.

(Oh, @bjorn3 had already answered: https://github.com/Rust-GCC/cargo-gccrs/pull/21#discussion_r665200941 Sorry!)

In order to differentiate, we can simply check the first argument. I agree that having multiple binaries make sense, and I think it's a better way to go about it and it's clearer.

>     * [ ]  `GccrsArgs` not holding all arguments
> 
> 
> https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L171
> 
> I still don't quite get why it is necessary to return a `Vec<GccrsArgs>`. Just from the struct name and the documentation, I would assume that when I have a `GccrsArgs` instance, that I have all the arguments that I need. This is also the main thing that prevents the `GccrsArgs` from implementing things like the `TryFrom` trait and so on.

For a single `rustc` invocation, we might need multiple `gccrs` invocation. For example, `rustc --crate-type=bin --crate-type=dyn --crate-type=static file.rs` would create a binary, a shared executable and a static library. This behavior is not reproducible in `gcc` as the compiler can only generate a single output file (as far as I know), so we need to call `gccrs` three times. Once for the binary, once for the shared library and once for the static library. Therefore, we need three different sets of arguments...

Note that this might change thanks to @bjorn3 's really nice and detailed proposition [here](https://github.com/Rust-GCC/cargo-gccrs/issues/34) that we must discuss.

>     * [ ]  `GccrsArgs` is doing too much
> 
> 
> https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L164-L168
> 
> I still think that with this, `GccrsArgs` is doing too much. One good principle of OOP is that one class should only do one thing. With that principle, `GccrsArgs` should only handle arguments, which includes parsing, converting, and then holding them. Now it handles arguments and does random other things like creating archives and in the future possibly even more.
> This is the area of the current implementation that really needs a refactor IMHO. Apply the KISS principle as often as you can. I have trouble following what is done where, so the current impl isn't really simple to follow.

Good point. I'll get to the refactoring of `GccrsArgs` really soon! I want to get most of the test harness done before I get on that, and then I'll work on making the code as good as it can be.

I agree that the impl is getting way too big and hard to follow. I haven't worked on it since adding the callback, and it's in desperate need of a refactoring.

**Comment by bjorn3** [comment]
`cfg!(release)` is always false unless you pass `--cfg release` to rustc. There is `debug_assertions`, but no general way to cfg code out depending on the cargo profile.

**Comment by CohenArthur** [comment]
> `cfg!(release)` is always false unless you pass `--cfg release` to rustc. There is `debug_assertions`, but no general way to cfg code out depending on the cargo profile.

Woops :( Well then I'll remove it. It made sense in theory haha

**Comment by flip1995** [comment]
Finished in #21 

Great work!
